### PR TITLE
gltfio: introduce MaterialProvider interface.

### PR DESCRIPTION
**Pull request by prideout**
This is a refactoring that prepares for an alternate "sans filamat" implementation of `MaterialProvider`, which will be especially useful for web.

### CODE REVIEW

File: libs/gltfio/include/gltfio/MaterialProvider.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+namespace gltfio {
+
+enum class AlphaMode : uint8_t {
+    OPAQUE,
+    MASK,
+    BLEND
+};
+
+// NOTE: This key is processed by MurmurHashFn so please make padding explicit.
+struct alignas(8) MaterialKey {
```
Comment (romainguy) [inline libs/gltfio/include/gltfio/MaterialProvider.h]: You should assert on a sizeof(MaterialKey), just to check that compilers don't add extra padding. We had to use a #pragma in other places before.

File: libs/gltfio/include/gltfio/MaterialProvider.h [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+ *
+ * - The "MaterialGenerator" implementation generates materials at run time (which can be slow)
+ *   and requires the filamat library, but produces streamlined shaders.
+ *
+ * - The "Ubermaterial" implementation uses a small number of pre-built materials with complex
+ *   fragment shaders, but does not require any run time work or usage of filamat.
+ */
+class MaterialProvider {
+public:
+    static MaterialProvider* createMaterialGenerator(filament::Engine* engine);
```
Comment (romainguy) [inline libs/gltfio/include/gltfio/MaterialProvider.h]: Change the name of the method

File: libs/gltfio/include/gltfio/MaterialProvider.h [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+    static MaterialProvider* createUbermaterial(filament::Engine* engine) {
```
Comment (romainguy) [inline libs/gltfio/include/gltfio/MaterialProvider.h]: Should this be public?

Comment (prideout) [reply libs/gltfio/include/gltfio/MaterialProvider.h]: Yes I think clients should be able to create instances of MaterialProvider.  For example, MeshAssimp (if we were to keep it) could use it to create materials.

File: libs/gltfio/include/gltfio/MaterialProvider.h (hunk 2 above)
Comment (prideout) [reply libs/gltfio/include/gltfio/MaterialProvider.h]: This naming is intentional: "createMaterialGenerator" creates a certain implementation of MaterialProvider (which is abstract).

File: libs/gltfio/include/gltfio/MaterialProvider.h (hunk 1 above)
Comment (prideout) [reply libs/gltfio/include/gltfio/MaterialProvider.h]: Good idea, added a static_assert.

**Pull request merged**